Merge wied03/savant_2.0.0

wied03 opened the pull request:
PR details

wied03 committed:
Update IMLs, use latest idea plugin, use def for clearer Groovy scoping

wied03 committed:
Groovy update

wied03 committed:
Use correct semantic version

wied03 committed:
Latest/dogfood

wied03 committed:
Latest dogfooding

wied03 committed:
Use our cache directory explicitly

Avoid building test IMLs that depend on where things are run from

wied03 committed:
Updated version for idea-plugin to 2.0.0

wied03 committed:
better cast

wied03 committed:
Correct our cache path and tests

wied03 committed:
Update this test too

wied03 committed:
Drop a Maven dependency in here

wied03 committed:
Replicate our problem

wied03 committed:
Ensure we don't waffle (our problem is now replicated)

wied03 committed:
Format

wied03 committed:
Better explain

wied03 committed:
Updated version for idea-plugin to 2.0.1

wied03 committed:
Updated version for idea-plugin to 2.0.0

wied03 committed:
DRY savant files

wied03 commented on src/main/groovy/org/savantbuild/plugin/dep/idea/IDEAPlugin.groovy:
```diff
@@ -139,12 +139,21 @@ class IDEAPlugin extends BaseGroovyPlugin {
 
   private toRelativePATH(Path path, String userHome) {
     def artifactRealPath = path.toRealPath().toString()
+    // If we have a cache process defined, getting the first one is useful because it gives a more precise
```
this was not part of the fix directly, but it did enable me to run the tests with different paths and not fail (and would also be useful in a GHA runner for the same purpose)

wied03 reviewed (commented):
self review

wied03 commented on src/test/resources/expected.iml:
```diff
@@ -52,7 +63,7 @@
       <library>
         <CLASSES>
-          <root url="jar://$USER_HOME$/dev/inversoft/savant/savant-dependency-management/test-deps/integration/org/savantbuild/test/integration-build/2.1.1-{integration}/integration-build-2.1.1-{integration}.jar!/" />
+          <root url="jar://$USER_HOME$/dev/savant_source/savant/savant-dependency-management/test-deps/integration/org/savantbuild/test/integration-build/2.1.1-{integration}/integration-build-2.1.1-{integration}.jar!/" />
```
This is trading 1 home directory dependent location for another. I should go back and address this, even though it's just a test file.

wied03 reviewed (commented):
self review

robotdan commented on src/main/groovy/org/savantbuild/plugin/dep/idea/IDEAPlugin.groovy:
```diff
@@ -139,12 +139,21 @@ class IDEAPlugin extends BaseGroovyPlugin {
     // - While unlikely, a path could repeat, and we only want to replace the prefix of the path
 
-    if (artifactRealPath.startsWith(projectRealPath)) {
+    if (cacheDir && artifactRealPath.startsWith(cacheDir.canonicalPath)) {
```
Looks ok to me, but it may be best to have Brian take a look as well. But it seems like the right thing to do.

voidmain replied:
This looks pretty odd to me. The plugins shouldn't need to know where the fetch and publish processes are putting things. They are all relative to the file and module/project location. If this is just for testing, I'd revert it and have the tests use the relative paths, even if they don't look the best.

wied03 replied:
The problem I am solving, with the impl change, was that previously, these tests would fail if you didn't clone the Savant code in exactly the same spot (since our test deps are in another repo (savant-dep-mgmt), the expected IMLs are user home dependent).

I changed the impl to find the cache dir, even if it was outside the project dir, which was 1 way of dealing with that. I just pushed (perhaps better way), which instead of modifying the impl, in the test, I just copy the deps from the savant-dep-mgmt repo to our project dir (`build/test`), which then ensures we don't fallback to `$USER_HOME$` with our test expecations.

voidmain commented on build.savant:
```diff
@@ -50,11 +51,12 @@ project(group: "org.savantbuild.plugin", name: "idea", version: "2.0.0-RC.7", li
-groovyTestNG = loadPlugin(id: "org.savantbuild.plugin:groovy-testng:2.0.0-RC.6")
-idea = loadPlugin(id: "org.savantbuild.plugin:idea:2.0.0-RC.7")
-release = loadPlugin(id: "org.savantbuild.plugin:release-git:2.0.0-RC.6")
+def dependency = loadPlugin(id: "org.savantbuild.plugin:dependency:2.0.0-RC.7")
```
Same comment as the `savant-dependency-management` repo regarding the use of `def`.

wied03 replied:
Reverted after talking w/ Daniel.

voidmain commented on build.savant:
```diff
@@ -64,12 +66,12 @@ groovyTestNG.settings.groovyVersion = "4.0"
-    "org.savantbuild:savant-utils:2.0.0-RC.6"                 : "savant-utils",
-    "org.savantbuild:savant-version:2.0.0-RC.6"               : "savant-version",
-    "org.savantbuild.plugin:dependency:2.0.0-RC.6"            : "dependency-plugin"
+    ("org.savantbuild:savant-core:${savantVersion}".toString())                 : "savant-core",
```
Same comment about using the version in the string.

wied03 replied:
Reverted after talking w/ Daniel.

voidmain reviewed (commented):
I'm not clear on what problem this is solving. Maybe we need to jump on a screen share so you can show me the issue.

wied03 committed:
Remove def prefixes on plugins

wied03 committed:
Go back to previous style of idea module map

wied03 committed:
no need to pin to groovy patch version

wied03 committed:
Make tests runnable from any directory using non-impl changes

Instead of adapting the plugin to replace directories outside the project dir (from savant-dependency-management), copy those into the project directory

robotdan commented on idea-plugin.iml:
```diff
@@ -11,50 +11,49 @@
     <orderEntry type="sourceFolder" forTests="false" />
-    <orderEntry type="library" name="groovy-4.0.5" level="application" />
-    <orderEntry type="library" name="Groovy 4.0" level="application" />
+    <orderEntry type="library" name="groovy-4.0" level="application" />
```
`Groovy 4.0`?

wied03 replied:
Sorry. missed. wrong convention. Fixed.

wied03 committed:
wrong convention
